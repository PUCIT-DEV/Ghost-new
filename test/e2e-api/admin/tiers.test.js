const {
    agentProvider,
    fixtureManager,
    mockManager,
    matchers
} = require('../../utils/e2e-framework');

describe('Tiers API', function () {
    let agent;

    before(async function () {
        agent = await agentProvider.getAdminAPIAgent();
        await fixtureManager.init('members');
        await agent.loginAsOwner();
    });

    beforeEach(function () {
        mockManager.mockLabsEnabled('multipleProducts');
    });

    afterEach(function () {
        mockManager.restore();
    });

    it('Can browse Tiers', async function () {
        await agent
            .get('/tiers/')
            .expectStatus(200)
            .matchBodySnapshot({
                tiers: Array(2).fill({
                    id: matchers.anyObjectId,
                    created_at: matchers.anyDate,
                    updated_at: matchers.anyDate
                })
            });
    });

    it('Errors when price is non-integer', async function () {
        const tier = {
            name: 'Blah',
            monthly_price: {
                amount: 99.99
            }
        };

        await agent
            .post('/tiers/')
            .body({tiers: [tier]})
            .expectStatus(422)
            .matchBodySnapshot({
                errors: [{
                    id: matchers.anyUuid
                }]
            });
    });

    it('Errors when price is negative', async function () {
        const tier = {
            name: 'Blah',
            monthly_price: {
                amount: -100
            }
        };

        await agent
            .post('/tiers/')
            .body({tiers: [tier]})
            .expectStatus(422)
            .matchBodySnapshot({
                errors: [{
                    id: matchers.anyUuid
                }]
            });
    });

    it('Errors when price is too large', async function () {
        const tier = {
            name: 'Blah',
            monthly_price: {
                amount: Number.MAX_SAFE_INTEGER
            }
        };

        await agent
            .post('/tiers/')
            .body({tiers: [tier]})
            .expectStatus(422)
            .matchBodySnapshot({
                errors: [{
                    id: matchers.anyUuid
                }]
            });
    });
});
