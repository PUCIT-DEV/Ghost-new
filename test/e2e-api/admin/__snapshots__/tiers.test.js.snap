// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tiers API Can browse Tiers 1: [body] 1`] = `
Object {
  "meta": Object {
    "pagination": Object {
      "limit": 15,
      "next": null,
      "page": 1,
      "pages": 1,
      "prev": null,
      "total": 2,
    },
  },
  "tiers": Array [
    Object {
      "active": true,
      "benefits": null,
      "created_at": "2022-02-21T16:04:06.000Z",
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Free",
      "slug": "free",
      "stripe_prices": null,
      "type": "free",
      "updated_at": "2022-02-21T16:04:09.000Z",
      "welcome_page_url": "/welcome-free",
    },
    Object {
      "active": true,
      "benefits": null,
      "created_at": "2022-02-21T16:04:06.000Z",
      "description": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "name": "Default Product",
      "slug": "default-product",
      "stripe_prices": null,
      "type": "paid",
      "updated_at": "2022-02-21T16:04:09.000Z",
      "welcome_page_url": "/welcome-paid",
    },
  ],
}
`;

exports[`Tiers API Errors when price is negative 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Tier prices must not be negative",
      "details": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Validation error, cannot save tier.",
      "property": null,
      "type": "ValidationError",
    },
  ],
}
`;

exports[`Tiers API Errors when price is non-integer 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Tier prices must be an integer.",
      "details": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Validation error, cannot save tier.",
      "property": null,
      "type": "ValidationError",
    },
  ],
}
`;

exports[`Tiers API Errors when price is too large 1: [body] 1`] = `
Object {
  "errors": Array [
    Object {
      "code": null,
      "context": "Tier prices may not exceed 999999.99",
      "details": null,
      "help": null,
      "id": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "message": "Validation error, cannot save tier.",
      "property": null,
      "type": "ValidationError",
    },
  ],
}
`;
